server/server/src/static_files.rs

```diff
@@ -70,3 +95,127 @@ async fn plugins(
 
     Ok(response)
 }
+
+async fn handle_file_upload(
+    mut payload: Multipart,
+    settings: Data<Settings>,
+    file_category: StaticFileCategory,
+) -> Result<Vec<UploadedFile>, Error> {
+    let service = StaticFileService::new(&settings.server.base_dir)
+        .map_err(|err| InternalError::new(err, StatusCode::INTERNAL_SERVER_ERROR))?;
+
+    let mut files = Vec::new();
+
+    while let Some(mut field) = payload.try_next().await? {
+        // A multipart/form-data stream has to contain `content_disposition`
+        let content_disposition = field.content_disposition();
+        log::info!(
+            "Uploading File: {}",
+            content_disposition.get_filename().unwrap_or_default()
+        );
+        log::debug!("Content Disposition: {:?}", content_disposition);
+        log::debug!("Content Type: {:?}", field.content_type());
+
+        let sanitized_filename =
+            sanitize_filename::sanitize(content_disposition.get_filename().unwrap_or_default());
+        let static_file = service
+            .reserve_file(&sanitized_filename, &file_category)
+            .map_err(|err| InternalError::new(err, StatusCode::INTERNAL_SERVER_ERROR))?;
+
+        files.push(UploadedFile {
+            id: static_file.id.clone(),
+            name: content_disposition
+                .get_name()
+                .unwrap_or_default()
+                .to_string(),
+            filename: content_disposition
+                .get_filename()
+                .unwrap_or_default()
+                .to_string(),
+            mime_type: field
+                .content_type()
+                .map(|mime| mime.to_string())
+                .unwrap_or_default(),
+        });
+
+        // File::create is blocking operation, use threadpool
+        let mut f = web::block(|| std::fs::File::create(static_file.path)).await??;
+
+        // Field in turn is stream of *Bytes* object
+        while let Some(chunk) = field.try_next().await? {
+            // filesystem operations are blocking, we have to use threadpool
+            f = web::block(move || f.write_all(&chunk).map(|_| f)).await??;
+        }
+    }
+    Ok(files)
+}
+
+async fn upload_sync_file(
+    payload: Multipart,
+    settings: Data<Settings>,
+    service_provider: Data<ServiceProvider>,
+    path: web::Path<(String, String)>,
+) -> Result<HttpResponse, Error> {
+    let db_connection = service_provider
+        .connection()
+        .map_err(|err| InternalError::new(err, StatusCode::INTERNAL_SERVER_ERROR))?;
+
+    let (table_name, record_id) = path.into_inner();
+
+    let files = handle_file_upload(
+        payload,
+        settings,
+        StaticFileCategory::SyncFile(table_name.clone(), record_id.clone()),
+    )
+    .await?;
+
+    let repo = SyncFileReferenceRowRepository::new(&db_connection);
+    for file in files.clone() {
```

Comment:
the response has `name: "files"` returned. What is this? seems hard-coded, so is it helpful?

<img width="565" alt="Screenshot 2024-03-20 at 11 14 42 AM" src="https://github.com/msupply-foundation/open-msupply/assets/9192912/1d2b7e50-ab8e-45f1-8082-1a0bc71868ac">

Reply:
![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/08b2cba3-e3d6-4353-95d7-47cbdd9b8226)